org.eclipse.lsp4e/src/org/eclipse/lsp4e/Versioned.java

```diff
@@ -0,0 +1,38 @@
+/*******************************************************************************
+ * Copyright (c) 2022-3 Cocotec Ltd and others.
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *  Ahmed Hussain (Cocotec Ltd) - initial implementation
+ *
+ *******************************************************************************/
+package org.eclipse.lsp4e;
+
+/**
+ * Bundles together a result from a language server with the document version at the time it was run.
+ * Supports optimistic locking.
+ *
+ * @param <T>
+ */
+public class Versioned<T> {
```

Comment:
I have remembered that I have remembered that org.eclipse.lsp4e.operations.semanticTokens.SemanticHighlightReconcilerStrategy also has checks to skip part of the work if the document is changed in between request and its response.
Shall we name this class VersionedDocumentOperation and have a constructor such as
`public Versioned(final IDocument document, final T data)`
and move `getDocumentModificationStamp` inside the class and make in private?